### ISSUE TITLE
wasm-merge emits non-imported global initializer expression

### ISSUE THREAD

**Issue by doehyunbaek** [WebAssembly/binaryen#6220]
index.wat
```wat
(module
  (import "env" "DYNAMICTOP_PTR" (global i32))
  (global (mut i32) (global.get 0))
  (func (export "_start") (;3480;)
    global.get 1
    drop
  )
)
```

```wat
(module
    (global (export "DYNAMICTOP_PTR") i32 (i32.const 9568))
)
```

```sh
~/binaryen-report/global_get_import wat2wasm index.wat && wat2wasm env.wat  && wasm-merge index.wasm index env.wasm env -o merged.wasm && wasm2wat merged.wasm
merged.wasm:0000019: error: global variable out of range: 1 (max 1)
```

It seems like in the resulting binary `merged.wasm`, global's initialization expression is still `global.get 0`, which is out of range. Reversing the two global declarations is also problematic with the error `initializer expression can only reference an imported global`.
```wat
  (global (;0;) (mut i32) (global.get 1))
  (global (;1;) i32 (i32.const 9568))
```

Interestingly, if you remove the usage of global 1 at `global.get 1`, wasm-merge successfuly merges the globals, as below

```wat
(module
  (type (;0;) (func))
  (func (;0;) (type 0))
  (global (;0;) i32 (i32.const 9568))
  (export "_start" (func 0))
  (export "DYNAMICTOP_PTR" (global 0)))
```

Maybe it's not a bug and out of scope for wasm-merge, but as I thought wasm-merge should give at least a warning on this case , I am reporting now.

Code is also at [this repo](https://github.com/doehyunbaek/binaryen-report/tree/master/global_get_import).

If this is a duplicate issue or non-issue, sorry in advance. Feel free to close the issue in that case.

I opened a separate issue with https://github.com/WebAssembly/binaryen/issues/6219 as there seems to be separate root cause, but if you feel like they should be consolidated please close this issue.

**Comment by kripken** [comment]
Thanks @doehyunbaek , this is indeed a bug. Fix in #6221

**Comment by doehyunbaek** [comment]
Hi @kripken. It seems that running the test case I mentioned, it still emits error `merged.wasm:0000020: error: initializer expression can only reference an imported global`, when run with `wasm-validate` is this intended?

Originally, emitted globals are like this
```
  (global (;0;) (mut i32) (global.get 1))
  (global (;1;) i32 (i32.const 9568))
```
After the fix, emitted globals are like this
```
  (global (;0;) i32 (i32.const 9568))
  (global (;1;) (mut i32) (global.get 0))
```

I am a little bit weird that this might be some quirkiness on `wasm-validate`'s part, as I could run with v8 even though wasm-validate errors with the above message. I found one github [issue](https://github.com/WebAssembly/wabt/issues/680) with regards to this in the wabt repo. Maybe this is something disallowed in spec but allowed in practice in engines?

I'm opening this again for this comment. You can close again if this is intended behavior.

**Comment by kripken** [comment]
That is expected. It will also not validate in `wasm-opt` by itself, but it will with `-all`. Likewise `--enable-all` should fix things in wabt, if it is up to date with the latest specs.

The reason is that
```wat
  (global (;0;) i32 (i32.const 9568))
  (global (;1;) (mut i32) (global.get 0))
```
is only valid if the GC extension is enabled. That's a little weird, but before that proposal a global initializer could only read from an *imported* global. With GC, that is relaxed so that reading a non-imported global works as well. (This was useful for GC so it was bundled into that proposal, but it by itself is not really related to GC.)

Here is the relevant validation code in Binaryen for reference:

https://github.com/WebAssembly/binaryen/blob/b4dee3dc05834ada8bf94e3e925186bc8b430c30/src/ir/properties.cpp#L65-L67

You can work around this by running `wasm-opt -O` or such. It should apply the first global's value to the second global, removing the `global.get` in the process. That may be necessary if you want to run in older VMs that do not support GC.

**Comment by doehyunbaek** [comment]
Oh thanks for sharing your knowledge.  I think wabt is not really up-to-date with the latest spec as it didn't fix with `--enable-all`.

Also huge thanks to pointing out the workaround, I am really impressed by your knowledge and willingness to share in such detail. Thank you.

**Comment by doehyunbaek** [comment]
Hi @kripken. Kinda worried that I keep bothering you with the expected semantics of binaryen. But I have one more question. Please forgive me if this is also an expected behavior.

> You can work around this by running wasm-opt -O or such. It should apply the first global's value to the second global, removing the global.get in the process. 

This optimization, presumably from here: 

https://github.com/WebAssembly/binaryen/blob/45f5bd2f440d6db5737947434cecdc4ffa37f567/src/passes/SimplifyGlobals.cpp#L668-L684

Doesn't apply to initializer expressions in element segment. For example,

```wat
// index.wat
(module
  (import "env" "DYNAMICTOP_PTR" (global i32))
  (table 1 funcref)
  (elem (;0;) (global.get 0) func 0)
  (func (export "entry") (param i32)
    (table.get 0 (i32.const 0))
    return
  )
)
```

```wat
// env.wat
(module
    (global (export "DYNAMICTOP_PTR") i32 (i32.const 0))
)
```

After wasm-merge and wasm-opt, it errors with `initializer expression can only reference an imported global`. I also have a code repo with this example [here](https://github.com/doehyunbaek/binaryen-report/tree/master/global_get_elem_init).

Is this an expected behavior?

If not, I tried a fix with this commit https://github.com/WebAssembly/binaryen/commit/01e26913186fc1c10fdfa7d80702e9a2cea6fcdb, and it at least happened to solve my use cases, including the example given above.

If this fix is appropriate, I would love to contribute a PR with a test case accompanied.

**Comment by kripken** [comment]
Thanks @doehyunbaek , good find! Yes, this was a missing optimization.

Your proposed fix is almost correct, but things are a little more complex I'm afraid. I almost wrote up a comment with a suggestion for a PR but I realized the comment would be as long as the code in the PR, so I just opened that instead here: #6226 . Thanks for offering to do the work, though!

(The specific reason I prefer the code in the PR is that the comment in the quoted code here is right that `propagateConstantsToGlobals` is the right place to do this optimization, since it can reason about the order of execution, specifically that even a mutable global can be applied in a segment offset, as mutation isn't possible there. The wasm spec actually doesn't allow mutable globals in offsets atm, but I suspect it may be extended in the future as it has already been in the GC proposal ([mentioned above](https://github.com/WebAssembly/binaryen/issues/6220#issuecomment-1888317914)), so the PR does things in a way that will optimize that case as well, if/when it arrives.)
